lib/core/test/unit/Cardano/Pool/DB/Properties.hs

```diff
@@ -1284,6 +1294,38 @@ prop_MultiPoolCertificateSequence_coverage mpcs = checkCoverage
     certificateSequences = getSinglePoolSequences mpcs
     poolCount = length certificateSequences
 
+
+-- | read . put == pure
+prop_putHeaderListHeader
+    :: DBLayer IO
+    -> [BlockHeader]
+    -> Positive Int
+    -> Property
+prop_putHeaderListHeader DBLayer{..} headers (Positive k) =
+    monadicIO (setup >> prop)
+  where
+    sortedHeaders = sortBy (\a b -> compare (blockHeight a) (blockHeight b))
+        . L.nubBy (\a b -> (blockHeight a) == (blockHeight b)) . nub $ headers
+    expected = reverse . (if k == 0 then id else take k) . reverse $ sortedHeaders
```

Comment:
I don't think we need the zero-value test here, as we have already specified that k is `Positive` (which QuickCheck [guarantees as being greater than `0`](http://hackage.haskell.org/package/QuickCheck-2.14.1/docs/Test-QuickCheck-Modifiers.html#t:Positive)).

But even if we did encounter a value of `0`, shouldn't we then just take `0` elements?

```suggestion
    expected = reverse . take k . reverse $ sortedHeaders
```

Reply:
> But even if we did encounter a value of 0, shouldn't we then just take 0 elements?

Nope, in that case we should take **all** elements. That's how `LimitTo` from persistent works. This emulates it.

Reply:
> Nope, in that case we should take all elements. That's how `LimitTo` from persistent works. This emulates it. And as such, we want the 0-case.

Is this documented anywhere?

I couldn't find any mention of this in [the Haddock documentation for `LimitTo`](https://hackage.haskell.org/package/persistent-2.10.5.2/docs/Database-Persist-Types.html#t:SelectOpt).

If this is the case, then we should definitely add a comment in the interface for `DBLayer.listHeaders`, to let callers know this is what they should expect.

Reply:
I don't know, but that's what the implementation does:

https://github.com/yesodweb/persistent/blob/master/persistent/Database/Persist.hs#L389-L398

https://github.com/yesodweb/persistent/blob/master/persistent/Database/Persist/Sql/Orphan/PersistQuery.hs#L396-L409

Reply:
> I don't know, but that's what the implementation does:

Thanks for confirming.

It's very surprising that this isn't documented anywhere! (I would have thought that the [`LimitTo` constructor](https://hackage.haskell.org/package/persistent-2.10.5.2/docs/Database-Persist-Types.html#t:SelectOpt) would be an excellent place to put this documentation.)

In any case, I think we should do one of the following:

1. Adjust the SQLite implementation of `listHeaders` to check for `limit <= 0` and then return `[]`.
2. Document the special significance of `limit <= 0` in the interface to `listHeaders`.

My preference would be for option 1, as I'm not a fan of using magic values. IMO, it should _never_ be the case that _reducing_ a limit causes a function to return _more_ values. This breaks the principle of monotonicity.

If OTOH we really need to support the case where caller wants all values, perhaps the limit can be specified as `Maybe Int`, or by using a sum type.

Reply:
I'd propose creating a separate ticket/PR for this. I added documentation about this behavior though.